File: Review-15-Rust-Strategy-Pattern/Example-4/src/game.rs
use std::fmt;

use crate::board::Board;
use crate::player::Player;
use crate::referee::Referee;

#[derive(Clone, Debug, Default)]
pub struct Player1NotSet;

#[derive(Clone, Debug, Default)]
pub struct Player2NotSet;

#[derive(Clone, Debug, Default)]
pub struct NotReady;

#[derive(Clone, Debug, Default)]
pub struct InProgress;

#[derive(Debug)]
pub enum EndState {
    Win,
    Stalemate,
}

#[derive(Clone, Debug, Default)]
pub struct Game<P1, P2, S> {
    player_1: P1,
    player_2: P2,
    board: Board,
    state: S
}

impl Game<Player1NotSet, Player2NotSet, NotReady> {
    pub fn new() -> Self {
        Game::default()
    }

    pub fn add_player<'game>(self, player: Player<'game>) -> Game<Player<'game>, Player2NotSet, NotReady> {
        Game {
            player_1: player,
            player_2: self.player_2,
            board: self.board,
            state: self.state
        }
    }
}

impl<'game> Game<Player<'game>, Player2NotSet, NotReady> {
    pub fn add_player(self, player: Player<'game>) -> Game<Player<'game>, Player<'game>, InProgress> {
        Game {
            player_1: self.player_1,
            player_2: player,
            board: Board::new(),
            state: InProgress
        }
    }
}

impl<'game> Game<Player<'game>, Player<'game>, InProgress> {
    fn do_one_turn(board: &mut Board, player: &mut Player, symbol: char) -> bool {
        println!("{}", board);
        println!();

        loop {
            let selected_move = player.next_move();

            if Referee::selected_cell_is_empty(selected_move, &board) {
                let _ = board.set_cell(selected_move, symbol);
                break;
            }
        }

        Referee::check_for_win(&board)
    }

    pub fn play_match(mut self) -> CompletedGame<'game> {
        while self.is_not_over() {
            if Self::do_one_turn(&mut self.board, &mut self.player_1, 'X') {
                println!("{}", self.board);
                println!();

                return CompletedGame {
                    winner: Some(self.player_1),
                    loser: Some(self.player_2),
                    end_state: EndState::Win,
                };
            }

            if Self::do_one_turn(&mut self.board, &mut self.player_2, 'O') {
                println!("{}", self.board);
                println!();

                return CompletedGame {
                    winner: Some(self.player_2),
                    loser: Some(self.player_1),
                    end_state: EndState::Win,
                };
            }
        }

        CompletedGame {
            winner: None,
            loser: None,
            end_state: EndState::Stalemate,
        }
    }

    pub fn is_over(&self) -> bool {
        false
    }

    pub fn is_not_over(&self) -> bool {
        !self.is_over()
    }
}

#[derive(Debug)]
pub struct CompletedGame<'game> {
    pub winner: Option<Player<'game>>,
    pub loser: Option<Player<'game>>,
    pub end_state: EndState,
}

impl<'game> CompletedGame<'game> {
    pub fn is_over(&self) -> bool {
        true
    }

    pub fn is_not_over(&self) -> bool {
        !self.is_over()
    }
}

impl<'game> fmt::Display for CompletedGame<'game> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.end_state {
            EndState::Win => {
                writeln!(
                    f,
                    "Congratulations {}!",
                    self.winner
                        .as_ref()
                        .expect("winner should be set...")
                        .get_name()
                )
            }
            EndState::Stalemate => {
                writeln!(f, "Stalemate...")
            }
        }
    }
}

File: Review-15-Python-Strategy-Pattern/Example-7/tictactoe/game.py
"""
This module handles the top-level game logic, including...

1. Player symbol assignment
2. Setting up the Referee
3. Tracking match progress
4. Managing player move order and validation
"""

from dataclasses import dataclass, field
from enum import StrEnum, auto

from .board import Board
from .factories import MoveStrategyFactory, RenderStrategyFactory
from .player import Player
from .referee import Referee


class GameStateError(Exception):
    """
    This exception is raised if a game is started before both players have been
    added and their symbols set.
    """


class CompletedGame:
    @dataclass(kw_only=True)
    class Win:
        board: Board = field(compare=True)

        winner: Player = field(compare=True)
        loser: Player = field(compare=True)

        def __str__(self) -> str:
            return f"Congratulations {self.winner.name}!"

    @dataclass(kw_only=True)
    class Stalemate:
        board: Board = field(compare=True)

        def __str__(self) -> str:
            return "Stalemate..."

    @dataclass(kw_only=True)
    class Forfeit:
        board: Board = field(compare=True)

        winner: Player = field(compare=True)
        loser: Player = field(compare=True)

        def __str__(self) -> str:
            return f"{self.loser.name} forfeited."


class TurnResult(StrEnum):
    WIN = auto()
    STALEMATE = auto()
    NOT_OVER_YET = auto()
    FORFEIT = auto()


@dataclass(kw_only=True)
class Game:
    """
    Orchestrates a single match of Tic-Tac-Toe.
    """

    board: Board = field(default_factory=Board)

    player1: Player
    player2: Player

    def __opponent_of(self, player: Player) -> Player:
        if player is self.player1:
            return self.player2

        return self.player1

    @staticmethod
    def use_defaults() -> None:
        MoveStrategyFactory.add_defaults()
        RenderStrategyFactory.add_defaults()

    def play_match(
        self,
    ) -> CompletedGame.Win | CompletedGame.Stalemate | CompletedGame.Forfeit:

        while True:
            players_and_symbols = ((self.player1, "X"), (self.player2, "O"))

            for player, symbol in players_and_symbols:
                player.get_render_preference().render(self.board)

                match self._player_turn(player, symbol):
                    case TurnResult.WIN:
                        return CompletedGame.Win(
                            board=self.board,
                            winner=player,
                            loser=self.__opponent_of(player),
                        )

                    case TurnResult.STALEMATE:
                        player.get_render_preference().render(self.board)

                        return CompletedGame.Stalemate(
                            board=self.board,
                        )

                    case TurnResult.FORFEIT:
                        raise NotImplementedError()

                    case _:
                        # the only possibility left is NOT_OVER_YET... no op
                        pass

    def _player_turn(self, player: Player, symbol: str) -> TurnResult:
        # Get a move and re-prompt if the move is invalid
        while True:
            move = player.next_move()

            if Referee.selected_cell_is_empty(self.board, move):
                break

        self.board.set_cell(move, symbol)

        if Referee.check_for_win(self.board):
            return TurnResult.WIN

        if self.board.is_full():
            return TurnResult.STALEMATE

        return TurnResult.NOT_OVER_YET
